lib/internal/assert/snapshot.js

```diff
@@ -0,0 +1,124 @@
+  SafeMap,
+  SafeSet,
+  StringPrototypeIndexOf,
+  StringPrototypeSplit,
+  StringPrototypeSubstring,
+  StringPrototypeReplace,
+  Symbol,
+} = primordials;
+
+const { emitExperimentalWarning } = require('internal/util');
+const { codes: { ERR_ASSERT_SNAPSHOT_NOT_SUPPORTED } } = require('internal/errors');
+const AssertionError = require('internal/assert/assertion_error');
+const { inspect } = require('internal/util/inspect');
+const { createReadStream, createWriteStream } = require('fs');
+const path = require('path');
+const assert = require('assert');
+
+const kUpdateSnapshot = Boolean(process.env.NODE_UPDATE_SNAPSHOT);
+const kInitialSnapshot = Symbol('kInitialSnapshot');
+const kDefaultDelimiter = '\n#*#*#*$#*#*#*#*#*#*$#*#*#\n';
+
+function defaultSnapshotPath() {
+  if (process.mainModule) {
+    const { dir, name } = path.parse(process.mainModule.filename);
+    return path.join(dir, `${name}.snapshot`);
+  }
+  if (!process.argv[1]) {
+    throw new ERR_ASSERT_SNAPSHOT_NOT_SUPPORTED();
+  }
+  const { dir, name } = path.parse(process.argv[1]);
+  return path.join(dir, `${name}.snapshot`);
+}
+
+let _source;
+function lazySource() {
+  _source = _source ?? createReadStream(defaultSnapshotPath(), { encoding: 'utf8' });
+  return _source;
+}
+
+let _target;
+function lazyTarget() {
+  _target = _target ?? createWriteStream(defaultSnapshotPath(), { encoding: 'utf8' });
+  return _target;
+}
+
+function serializeName(name) {
+  return StringPrototypeReplace(`${name}`, /:\n/g, '_');
+}
+
+let writtenNames;
+let snapshotValue;
+let counter;
+
+function writeSnapshot({ name, value }) {
+  const target = lazyTarget();
+  if (counter > 1) {
+    target.write(kDefaultDelimiter);
+  }
+  writtenNames = writtenNames || new SafeSet();
+  if (writtenNames.has(name)) {
+    throw new AssertionError({ message: `Snapshot "${name}" already used` });
+  }
+  writtenNames.add(name);
+  target.write(`${name}:\n${value}`);
+}
+
+async function getSnapshot() {
+  const source = lazySource();
+  if (snapshotValue !== undefined) {
+    return snapshotValue;
+  }
+  if (kUpdateSnapshot) {
+    snapshotValue = kInitialSnapshot;
+    return kInitialSnapshot;
+  }
+  try {
+    let data = '';
+    for await (const line of source) {
+      data += line;
+    }
+    snapshotValue = new SafeMap(
+      ArrayPrototypeMap(
+        StringPrototypeSplit(data, kDefaultDelimiter),
+        (item) => {
+          const keyDelimiter = StringPrototypeIndexOf(item, ':\n');
+          return [StringPrototypeSubstring(item, 0, keyDelimiter), StringPrototypeSubstring(item, keyDelimiter + 2)];
+        }
+      ));
+  } catch (e) {
+    if (e.code === 'ENOENT') {
+      snapshotValue = kInitialSnapshot;
+    } else {
+      throw e;
+    }
+  }
+  return snapshotValue;
+}
+
+
+async function snapshot(input, name) {
+  emitExperimentalWarning('Snapshot');
+
+  const snapshot = await getSnapshot();
+  counter = (counter ?? 0) + 1;
+  name = serializeName(name ?? counter);
```

Comment:
q:  is it intentional the counter incremented even if `name` is used?

Reply:
yes, so we know when a delimiter is needed